Create the SmartAds package

michalczaplinski opened the pull request:
**What**:

Create the new `@frontity/smart-adserver` package. 

Following the Feature Discussion: https://community.frontity.org/t/smart-adserver/1586

**Why**:

To integrate Smart Adserver into Frontity.

**How**:

- Create the new `@frontity/smart-adserver` package in the frontity-monorepo
- Expose the root which will include the `<Head />` that loads the Smart Adserver library
- Allow the user to pass settings to the package using `frontity.settings.js`
- Create the `SmartAd` component that will be used in the `libraries`.

**Tasks**:

- [x] Code
- [x] TSDocs
- [x] TypeScript
- [x] Unit tests
- [x] End to end tests
- [x] Open an issue for this feature in the [docs repo](https://github.com/frontity/docs/wiki/Code-Releases): https://github.com/frontity/docs/issues/234
- [x] Update community discussions: https://community.frontity.org/t/smart-adserver/1586/10
- [x] Add a changeset (with link to its [Feature Discussion](https://community.frontity.org/c/33) if it exists)

<!-- Changesets are necessary if your changes should release any packages.
Run `npx changeset` to create a changeset.
More info at https://docs.frontity.org/contributing/code-contribution-guide#what-is-a-changeset -->

**Unrelated Tasks**

<!-- ignore-task-list-start -->
- [x] TypeScript tests
- [x] Update starter themes
- [x] Update other packages
<!-- ignore-task-list-end -->

michalczaplinski committed:
Create the smart-adserver package

michalczaplinski committed:
Add changeset

michalczaplinski committed:
Merge branch 'dev' into smart-ad-server

michalczaplinski committed:
Add the e2e tests for smart-adserver (WIP)

Commit:
Remove unused dependencies and fix types

Commit:
Add types for networkid & subdomain

Commit:
Add the smart-ad component and fix types

Commit:
Add smartserver e2e test

Commit:
Fix the @frontity/smart-adserver version

DAreRodz commented:
> **Additional Comments**
>
> :warning: The `SmartAd` component will be implemented in another PR.

Are you sure? It looks like you [already implemented it](https://github.com/frontity/frontity/pull/569/files#diff-6d9700a19725c6d1912655373318ef4c). :joy:

michalczaplinski commented:
> > **Additional Comments**
> > ⚠️ The `SmartAd` component will be implemented in another PR.
> 
> Are you sure? It looks like you [already implemented it](https://github.com/frontity/frontity/pull/569/files#diff-6d9700a19725c6d1912655373318ef4c). 😂

I think it'll make more sense to implement them together so that I can just test them with the e2e tests 🙂

michalczaplinski commented:
<img width="855" alt="Screen Shot 2020-09-29 at 9 46 26 AM" src="https://user-images.githubusercontent.com/5417266/94574363-a6839f00-0238-11eb-893d-240d42e6a955.png">

new laptop, I've realized that I didn't set my git email, whoops

Commit:
Prevent imports from name collisions

Commit:
Add the dependencies for the smart adserver

Commit:
Merge branch 'dev' of github.com:frontity/frontity into smart-ad-server

michalczaplinski committed:
Update the test to TS, fix types, add dummy test cases

michalczaplinski committed:
remove the scirpt from useEffect

michalczaplinski committed:
Add types for CallOptions

michalczaplinski committed:
Add react "rules of hooks" to eslint

michalczaplinski committed:
Change the domain to www3

michalczaplinski committed:
Add TSDocs to all the types

michalczaplinski committed:
Make the smart ads work by storing reference to window.sas in frontity state.

michalczaplinski committed:
Add support for various call types

michalczaplinski committed:
Add more test stubs

michalczaplinski committed:
Add more comments and warnings if networkId or subdomain are not defined

michalczaplinski committed:
Merge branch 'dev' of github.com:frontity/frontity into smart-ad-server

michalczaplinski committed:
Add more test stubs

michalczaplinski committed:
Remove the context from the smart ad

michalczaplinski committed:
Fix the styles for the smart ad component

michalczaplinski committed:
Change .sas to .isLoaded

michalczaplinski committed:
Move the connect() to make the libraries work

michalczaplinski committed:
Add proper TSDocs and props for the SmartAd component

michalczaplinski committed:
Add a fallback for tagId

michalczaplinski committed:
Create more test cases for smart ads

michalczaplinski committed:
Add a super basic unit test

michalczaplinski committed:
export Smart Ad without connect and fix test case

michalczaplinski commented on .eslintrc:
```diff
@@ -8,8 +8,10 @@
     "plugin:cypress/recommended"
   ],
-  "plugins": ["jsdoc"],
+  "plugins": ["jsdoc", "react-hooks"],
```
This is the official eslint plugin from the react team: https://www.npmjs.com/package/eslint-plugin-react-hooks

michalczaplinski committed:
Merge branch 'dev' of github.com:frontity/frontity into smart-ad-server

michalczaplinski committed:
Add back erroneusly removed jest types

michalczaplinski committed:
Update the README

michalczaplinski committed:
Update the tests to look for iframes instead of script elements

michalczaplinski committed:
Remove the name from the e2e package

michalczaplinski committed:
Remove the siteId & pageId & formatId

michalczaplinski committed:
Add a test case for unmounting the component and rendering another in CSR

michalczaplinski committed:
Comment out the test case for iframe until we have new pageId, formatId & siteId

michalczaplinski committed:
Just fix the formatting

michalczaplinski committed:
Reformat the TSDoc slightly

michalczaplinski committed:
Reformat the TSDocs

luisherranz committed:
Merge branch 'dev' into smart-ad-server

luisherranz commented on packages/smart-adserver/src/components/smart-ad.tsx:
```diff
@@ -0,0 +1,70 @@
+        width: ${width}px;
+        height: ${height}px;
+      `
+    : css``;
```
Can't we also style the div of `std`?

Ads are always a mess for Cumulative Layout Shift (https://web.dev/cls/) so I remember that in the past we always tried to add a `height`.

Maybe the `height` value can be `height` for the iframe and `min-height` for the `div`, so if a bigger ad is loaded it's not cropped.

@darerodz, what do you think?

DAreRodz replied:
I agree. :+1:
I think we did something similar in the past.

luisherranz committed:
Simplify E2E tests

luisherranz committed:
Switch to major release

luisherranz reviewed (commented):
Great work Michal!

I didn't approve it because I'd prefer @darerodz to do so. He has more experience with ads than me, and he worked on our old smartads package and the current `@frontity/google-ad-manager` package.

Other than that, I've simplified the E2E tests and switched to a `major` bump for the 1.0.0 release.

I saw there are some Eslint warnings left too. I think they are coming from the new `react-hooks` eslint package.

luisherranz committed:
Fix package dep version in e2e tests

luisherranz committed:
TinyRouter: Check if `actions.source.fetch` exists

michalczaplinski committed:
Add the recommended useEffect dependencies

michalczaplinski committed:
Multiple changes:
- Add the minHeight
- Spread extra props in the SmartAd component
- Use useConnect to pass state to SmartAd component
- Pass an additional css prop to SmartAd component

michalczaplinski committed:
Add test case for minHeight

michalczaplinski commented:
@DAreRodz You're welcome to review this. Some notes for you:

1. I've allowed users to pass a `css` prop to add additional styles to the ad container as well as pass any other props to the container.

2. I've actually added all the dependencies suggested by the react-hooks eslint plugin because I've realized that it shouldn't be any problem. So, no eslint warnings anymore.

3. The `minHeight` prop on the `SmartAd` component on the does not have any default value, so the user has to pass it themselves, otherwise no height is specified.

DAreRodz reviewed (approved):
Everything all right. Great job @michalczaplinski!

PS: I just fixed two minor typos.

DAreRodz committed:
Fix minor typos
